Sources/Core/TimelineLoaderDecorator.swift

```diff
@@ -0,0 +1,256 @@
+        default: return true
+        }
+      }
+    default:
+      return true
+    }
+  }
+
+  fileprivate func nextPage() -> Int {
+    return sections.reduce(0, { $0 + $1.cells.count }) + 1
+  }
+
+  fileprivate func update(with items: [Item]?) {
+    switch state {
+    case .loading:
+      guard let items = items else { return }
+      guard let loader = loader else { return }
+
+      var mergedItems = self.items.filter { !items.contains($0) }
+      mergedItems.append(contentsOf: items)
+      mergedItems.sort()
+      self.items = mergedItems
+      sections = loader.sections(for: loader.cells(for: mergedItems))
+      registerCellsForSections()
+      containerView?.reloadData()
+    default:
+      assertionFailure("Should not be called in other state than loading")
+    }
+  }
+
+  // swiftlint:disable function_body_length
+  // swiftlint:disable:next cyclomatic_complexity
+  fileprivate func load(_ intent: LoaderIntent) {
+    guard DispatchQueue.isMain else {
+      return DispatchQueue.main.async {
+        self.load(intent)
+      }
+    }
+
+    guard needToHandleIntent(intent: intent) else { return }
+
+    if intent == .force(keepData: false) {
+      sections = []
+      reloadDataWithEmptyDataSet()
+    }
+
+    guard let itemsObservable = loader?.performLoading(intent: intent) else { return }
+    let cellsCountBeforeLoad = cellsCount
+    startProgress?(intent)
+    state = .loading(intent: intent)
+    let loaderDisposeBag = DisposeBag()
+    itemsObservable
+      .observeOn(MainScheduler.instance)
+      .subscribe(onNext: { [weak self] items in
+        self?.update(with: items)
+      }, onError: { [weak self] error in
+        guard let `self` = self else { return }
+        self.stopProgress?(intent)
+
+        self.state = self.cellsCount > 0 ? .hasData : .error(error)
+        self.reloadDataWithEmptyDataSet()
+      }, onCompleted: { [weak self] in
+        guard let `self` = self else { return }
+        self.stopProgress?(intent)
+        let cellsCountAfterLoad = self.cellsCount
+
+        switch intent {
+        case .page:
+          if cellsCountAfterLoad == cellsCountBeforeLoad {
+            self.state = .hasData
+            return
+          }
+        default: break
+        }
+        guard let containerView = self.containerView else { return }
+        if cellsCountAfterLoad == 0 {
+          self.state = .empty
+          self.reloadDataWithEmptyDataSet()
+        } else {
+          self.state = .hasData
+          guard let lastSection = self.sections.last else { return }
+          guard let visibleItems = containerView.visibleItems else { return }
+          let sectionsLastIndex = self.sections.count - 1
+          let itemsLastIndex = lastSection.cells.count - 1
+
+          if visibleItems.contains(where: { $0.section == sectionsLastIndex && $0.item == itemsLastIndex }) {
+            self.handleLastCellDisplayed()
+          }
+        }
+      }).disposed(by: loaderDisposeBag)
```

Comment:
Please move `self.stopProgress?(intent)` to onDisposed closure.

Reply:
yes, my lord